pkg/auth/providers/keycloakoidc/keycloak_provider.go

```diff
@@ -59,6 +59,12 @@ func newClient(config *v32.OIDCConfig, token *oauth2.Token) (*KeyCloakClient, er
 		return nil, err
 	}
 	oauthConfig := oidc.ConfigToOauthConfig(provider.Endpoint(), config)
+	// Valid will return false if access token is expired
+	if !token.Valid() {
+		// since token is not valid, the TokenSource func used in the Client func will attempt to refresh the access token
+		// if the refresh token has not expired
+		logrus.Debugf("[keycloak oidc] newClient: attempting to refresh access token")
+	}
```

Comment:
Where does the token get refreshed? Does the OIDC library handle that internally?

Reply:
Anytime TokenSource is used, its part of the oauth2 library to perform a refresh if needed. So after I log the message, there is use of the func TokenSource.  I am open to suggestions on a better way to determine refresh! 

https://github.com/rancher/rancher/blob/0952b35cc2b86d72d432f4834b2fa8ca1dbe85d8/pkg/auth/providers/oidc/oidc_provider.go#L384
https://github.com/golang/oauth2/blob/2e8d9340160224d36fd555eaf8837240a7e239a7/clientcredentials/clientcredentials.go#L71

https://github.com/rancher/rancher/blob/0952b35cc2b86d72d432f4834b2fa8ca1dbe85d8/pkg/auth/providers/keycloakoidc/keycloak_provider.go#L63
https://github.com/golang/oauth2/blob/2e8d9340160224d36fd555eaf8837240a7e239a7/clientcredentials/clientcredentials.go#L60

Reply:
Does that mean the token we carry if expired get refreshed internally, but we never get the refreshed token back? So we keep sending back the expired token?

Reply:
https://github.com/rancher/rancher/blob/edc77f8e255087257987d0920a84430ebb61da91/pkg/auth/providers/googleoauth/goauth_provider.go#L406 it looks similar to what we are doing here in the google provider. But you are correct, this means that once the initial token expires, every time a call is made, a refresh will occur.

Reply:
I wonder if we can call first to get refreshed token https://github.com/golang/oauth2/blob/2e8d9340160224d36fd555eaf8837240a7e239a7/oauth2.go#L298
If a different token is returned, we update it on our side as well as log that token refreshed message

Reply:
Going to make a separate issue for this so we can fix that later. 
https://github.com/rancher/rancher/issues/33916